Background image support

d-bingham opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
Initial code check-in for background image support.  Handles background images themselves, and adds options to profiles.json to control the feature.

<!-- Other than the issue solved, is this relevant to any other issues/existing PRs? --> 
## References

#833 

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [ ] Closes #833 (Not _quite_ complete as the opacity setting is not fully robust, see below)
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [x] Requires documentation to be updated -- Adds (currently undocumented) settings to profiles.json, but profiles.json as a whole does not appear to be documented
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #833 -- aware of the discussion in #833 

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

This adds three options to the individual profile options in profile.json:

- backgroundImage: Expects a Uri which locates an image.  This property is optional and defaults to null.
- backgroundImageOpacity: Expects a number, controls opacity of the background image.  Optional and defaults to 1.0.
- backgroundImageStretchMode: expects one of "none", "fill", "uniform", or "uniformToFill", which controls the stretch mode used -- see [documentation here](https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.stretch?view=netframework-4.8).  Optional and defaults to uniformToFill.

Background images are only displayed if useAcrylic is turned off.

One **known** issue:

backgroundImageOpacity currently only works with respect to the application's default black background and does not respect the profile's background color.

One _potential_ issue:

There's no attempt to handle a failed image load from the Uri; I _think_ it is reasonable to expect the underlying classes (specifically [BitmapImage](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.imaging.bitmapimage)) to handle this gracefully.  Additionally, it's unclear how such a failure should be reported to the user (beyond their requested image being displayed as a black console background instead)

zadjii-msft commented:
> Additionally, it's unclear how such a failure should be reported to the user

I am working on an error dialog in #622, but that's not quite ready yet, so I'm fine with this just doing whatever for now.

zadjii-msft commented on src/cascadia/TerminalControl/TermControl.cpp:
```diff
@@ -156,8 +156,12 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
     //   the settings are reloaded.
-    //   * Sets up the background of the control with the provided BG color,
-    //      acrylic or not, and if acrylic, then uses the opacity from _settings.
+	//   * Prioritizes the acrylic background if chosen, respecting acrylicOpacity
```
These comment lines look indented a little off

d-bingham replied:
Oops, missed these on the last commit.  Fixed.

zadjii-msft commented on src/cascadia/TerminalControl/TermControl.cpp:
```diff
@@ -189,6 +193,27 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+            Windows::Foundation::Uri imageUri{ _settings.BackgroundImage() };
+            Media::Imaging::BitmapImage image(imageUri);
+
+			// Note that BitmapImage handles the image load asynchronously,
```
There might be tabs here, looks a little too indented

d-bingham replied:
Fixed in the latest commit.

zadjii-msft commented on src/cascadia/TerminalSettings/IControlSettings.idl:
```diff
@@ -12,6 +12,14 @@ namespace Microsoft.Terminal.Settings
         Hidden
     };
 
+    enum ImageStretchMode
```
Do you need to create your own enum here? could `BackgroundImageStretchMode` be of type `...Media.Strech`? Then you wouldn't need to static_cast on TermControl.cpp:208 either

d-bingham replied:
There's no compelling reason to do so, I suppose; I was following the precedent in the Profile code, but in retrospect utilizing the existing enum when the intent is to exhaustively implement all values seems the more reasonable option.

Switched over to the existing enum on the most-recent commit.

DHowett-MSFT commented on src/cascadia/TerminalControl/TermControl.cpp:
```diff
@@ -189,6 +193,27 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+            brush.Stretch(static_cast<Media::Stretch>(_settings.BackgroundImageStretchMode()));
+            brush.Opacity(_settings.BackgroundImageOpacity());
+
+            _root.Background(brush);
```
If we introduce another layer just to hold the image, behind the control but within the content root, we probably get both acrylic and background image together for “free”!

d-bingham replied:
This seems like a good next step, yes.

dlong11 commented on src/cascadia/TerminalApp/Profile.cpp:
```diff
@@ -71,7 +80,10 @@ Profile::Profile() :
+    _icon{ },
+    _backgroundImage{ },
+    _backgroundImageOpacity{ },
+    _backgroundImageStretchMode{ }
```
@d-bingham It is unnecessary to initialize the added _backgroundXXXX methods here.   They are already initialized to the "empty state" when they are declared in the header.

d-bingham replied:
Strictly speaking it's not necessary, and style guidelines differ on whether to include them or not.  They're in there following the precedent in the rest of the profile code which explicitly initializes members with default constructors.

dlong11 replied:
@d-bingham you are correct that the style guides differ!  @binarycrusader what is the Windows Terminal's team thoughts on this?  I ping'd you because I _think_ you may have mentioned this in reviews.   Knowing what the expectations are would be helpful when reviewing.  (ie - if this is cool then we can stop looking for it)

binarycrusader replied:
I’m not on the Terminal team, merely an interested party that contributes.  However, I suspect the general idea is that new code generally conforms to the C++ guidelines and modifications to existing try to match existing style.

DHowett-MSFT replied:
@binarycrusader's quite correct. There's a balance we're still trying to strike here between what is _correct_ and what is _complete_ :smile: Thanks for bearing with us!

dlong11 replied:
@DHowett-MSFT No worries.   I am trying to find the correct balance when reviewing the code.  I don't want to mention stuff like this if it is not appropriate.    It appears the recommendations are being worked on, so all is good.    In the meantime - what is your opinion on commenting on stuff like this?  Should I hold off or continue to point these out?

DHowett-MSFT replied:
I think it's helpful for you to point these things out, but it's even more helpful if we start to accumulate `Area-CodeHealth` issues for things like it :smile:

dlong11 replied:
@DHowett-MSFT Sounds like a plan.

miniksa commented:
I'm concerned as to whether this should be a property of the `DxRenderer` itself instead of composed on during the XAML process in the control.

But given that I don't currently have the capacity to figure that out, I'm OK with accepting this in the meantime.

DHowett-MSFT commented:
> 
> 
> I'm concerned as to whether this should be a property of the `DxRenderer` itself instead of composed on during the XAML process in the control.
> 
> But given that I don't currently have the capacity to figure that out, I'm OK with accepting this in the meantime.

Yeah, we can stick with this. Putting it behind the `DxRenderer` (literally behind) frees us from thinking about image loading, format handling, blitting, invalidation, stretch, interpolation and everything else XAML gives us for ~free.

Given that list, I'm inclined to prefer the XAML solution. :smile:

d-bingham commented:
> Given that list, I'm inclined to prefer the XAML solution. 😄

Additionally, as far as I'm aware there's no "official" implementation of the Fluent acrylic brush in Direct2D, so the XAML layer would still be required for that effect.

d-bingham commented:
Working on cleaning up some architectural oddities created by merging this with #891.  The current PR build works, but it's inelegant (and the code gets called three times on a settings file change...)

d-bingham commented:
> @d-bingham you might not need to worry about the settings reloading issue too much. I know that it's definitely called twice per settings change, because for whatever reason, we get _two_ file change notifications.

I'm more concerned about getting a background image refresh (and the resulting flicker of the background image) resulting from _terminal_ input introduced in #891.

d-bingham commented:
Modified this a bit to make more sense with regard to #891:

- Split off brush-initialization work from #891's TermControl::_BackgroundColorChanged; _BackgroundColorChanged now only, er, changes background color
- Put the brush initialization into TermControl::_InitializeBackgroundBrush
- Enhanced TermControl::_InitializeBackgroundBrush so that the acrylic and image brushes avoid redraws (image refresh for the image brush, acrylic redraw and fade for acrylic) when they are reset

d-bingham commented:
Also, _critically_, the flicker prevention code maintains animation state of any animated image selected as a background during a settings change that does not change the image source.

zadjii-msft commented:
> animation state of any animated image selected as a background

Are you suggesting we can have [gifs as a background](https://gfycat.com/dopeygrimyhyena.gif)? Because that would be amazing

d-bingham commented:
> > animation state of any animated image selected as a background
> 
> Are you suggesting we can have [gifs as a background](https://gfycat.com/dopeygrimyhyena.gif)? Because that would be amazing


Yeah, just set your backgroundImage to https://lh6.googleusercontent.com/-Kde1FiWiijM/TniZsFpCyqI/AAAAAAAAD7Q/2sZC2Zqx8RM/s500/Contact500_Micael_Reynaud.gif and give it a bit to load, and it works.
